Framework/API/inc/MantidAPI/WorkspacePropertyWithIndex.tcc

```diff
@@ -0,0 +1,190 @@
+operator=(const std::tuple<boost::shared_ptr<TYPE>, API::IndexType,
+                           std::vector<int>> &rhs) {
+  boost::shared_ptr<TYPE> ws;
+  API::IndexType type;
+  std::vector<int> list;
+
+  std::tie(ws, type, list) = rhs;
+
+  *this = ws;
+  *this->m_indexTypeProp.get() = type;
+  *this->m_indexListProp.get() = list;
+
+  return *this;
+}
+
+template <typename TYPE>
+WorkspacePropertyWithIndex<TYPE> &WorkspacePropertyWithIndex<TYPE>::
+operator=(const std::tuple<boost::shared_ptr<TYPE>, API::IndexType, std::string>
+              &rhs) {
+  boost::shared_ptr<TYPE> ws;
+  API::IndexType type;
+  std::string list;
+
+  std::tie(ws, type, list) = rhs;
+
+  *this = ws;
+  *this->m_indexTypeProp.get() = type;
+  m_indexListProp->setValue(list);
+
+  return *this;
+}
+
+template <typename TYPE>
+WorkspacePropertyWithIndex<TYPE> &WorkspacePropertyWithIndex<TYPE>::
+operator=(const WorkspacePropertyWithIndex<TYPE> &rhs) {
+  if (&rhs == this)
+    return *this;
+  API::WorkspaceProperty<TYPE>::operator=(rhs);
+  *m_indexListProp = *rhs.m_indexListProp;
+  *m_indexTypeProp = rhs.m_indexTypeProp->selectedType();
+
+  return *this;
+}
+
+template <typename TYPE>
+WorkspacePropertyWithIndex<TYPE> &WorkspacePropertyWithIndex<TYPE>::
+operator+=(WorkspacePropertyWithIndex const *rhs) {
+  throw Kernel::Exception::NotImplementedError(
+      "+= operator is not implemented for WorkspacePropertyWithIndex.");
+  return *this;
+}
+
+template <typename TYPE>
+WorkspacePropertyWithIndex<TYPE> *
+WorkspacePropertyWithIndex<TYPE>::clone() const {
+  return new WorkspacePropertyWithIndex<TYPE>(*this);
+}
+
+template <typename TYPE>
+WorkspacePropertyWithIndex<TYPE>::
+operator const std::tuple<boost::shared_ptr<TYPE>, SpectrumIndexSet>() const {
+  return std::make_pair(operator()(), getIndices());
+}
+
+template <typename TYPE>
+WorkspacePropertyWithIndex<TYPE>::
+operator const std::tuple<boost::shared_ptr<const TYPE>, SpectrumIndexSet>()
+    const {
+  return std::make_pair(operator()(), getIndices());
+}
+
+template <typename TYPE>
+const SpectrumIndexSet WorkspacePropertyWithIndex<TYPE>::getIndices() const {
+  // TODO handle DetectorID->SpectrumNumber conversion
+  // This will need to be an extra IndexType
+  const auto &list = this->m_indexListProp->operator()();
+
+  const auto &indexInfo =
+      boost::dynamic_pointer_cast<API::MatrixWorkspace>(getWorkspace())
+          ->indexInfo();
+
+  // If no indices provided, then assume all
+  if (list.size() == 0)
+    return indexInfo.makeIndexSet();
+
+  switch (m_indexTypeProp->selectedType()) {
+  case IndexType::SpectrumNumber:
+    return indexInfo.makeIndexSet(std::vector<Mantid::Indexing::SpectrumNumber>(
+        list.begin(), list.end()));
+    break;
+  case IndexType::WorkspaceIndex:
+    return indexInfo.makeIndexSet(
+        std::vector<Mantid::Indexing::GlobalSpectrumIndex>(list.begin(),
+                                                           list.end()));
+    break;
+  }
```

Comment:
@SimonHeybrock  The list property (see ArrayProperty) converts the range to a vector by default. I could check the min and max values and see if the length of the list is the same as `max - min`.  If it is a pure range, then I would just use min and max, otherwise I will create the vector.